### PR TITLE
Flypeople now eat vomit

### PR DESCRIPTION
**Pull request by LanCartwright**
:cl: 
rscadd: Flypeople now vomit when they eat nutriment.
rscadd: Flypeople can now suck the vomit off the floor to gain it's nutritional content.
rscadd: Flypeople now must suck puke chunks and vomit to fill themselves of food content.
/:cl:

Flies can no longer eat.
Flies must eat vomit.
Eating causes vomiting.
Janitors and Janitorborgs rollout.

Edit:
Ported from my old server. hatestation13.
Original author: Koriath.
Assitance from Octareenroon91.
Massive support and assistance from oranges.

### CODE REVIEW

**Comment by YotaXP** [comment]
Woo immersion!

**Comment by TechnoAlchemisto** [comment]
so eating causes vomiting for flies? or eating for anyone causes vomiting?

**Comment by LanCartwright** [comment]
Flypeople only.

File: code/game/atoms.dm [hunk 1]
```diff
@@ -280,6 +280,13 @@ var/list/blood_splatter_icons = list()
 		var/obj/effect/decal/cleanable/vomit/this = new /obj/effect/decal/cleanable/vomit(src)
 		if(M.reagents)
 			M.reagents.trans_to(this, M.reagents.total_volume / 10)
+			for(var/datum/reagent/R in reagents.reagent_list)		//clears the stomach of anything that might be digested as food
```
Comment (optimumtact) [inline code/game/atoms.dm]: can you write this into a proc? nesting like this triggers me, so I'd like it locked away where I don't have to see it.

File: code/game/objects/effects/decals/Cleanable/misc.dm [hunk 2]
```diff
@@ -87,6 +87,22 @@
...
 	var/list/viruses = list()
 
+/obj/effect/decal/cleanable/vomit/attack_hand(var/mob/user)
+	if(istype(user,/mob/living/carbon/human))
+		var/mob/living/carbon/human/H = user
+		if(H.dna.species.id == "fly")
+			playsound(get_turf(src), 'sound/items/drink.ogg', 50, 1) //slurp
+			H.visible_message("<span class='alert'>[H] extends a small proboscis into the vomit pool, sucking it with a slurping sound.</span>")
+			if(reagents)
+				for(var/datum/reagent/R in reagents.reagent_list)
```
Comment (optimumtact) [inline code/game/objects/effects/decals/Cleanable/misc.dm]: plus then you can reuse the proc here

File: code/game/atoms.dm (hunk 1 above)
Comment (LanCartwright) [inline code/game/atoms.dm]: I'll have to learn to. This was something I ported.

**Comment by Zilenan91** [comment]
So how do they get the vomit if they need to eat, that stuff usually gets cleaned up in 2 seconds

**Comment by LanCartwright** [comment]
@Zilenan91 They can create vomit by eating.

**Comment by Zilenan91** [comment]
But if they can only eat vomit then how can they eat if there is no vomit

**Comment by LanCartwright** [comment]
They do not gain anything by eating vomit. Attempting to eat something creates a pool of vomit they can eat.

File: code/modules/mob/living/carbon/human/species_types.dm [hunk 3]
```diff
@@ -373,6 +373,21 @@ var/regex/lizard_hiSS = new("S+", "g")
 /datum/species/fly/handle_speech(message)
 	return replacetext(message, "z", stutter("zz"))
 
+/datum/species/fly/handle_chemicals(datum/reagent/chem, mob/living/carbon/human/H)
+	if(istype(chem,/datum/reagent/consumable))
+		var/datum/reagent/consumable/nutri_check = chem
+		if(nutri_check.nutriment_factor >0)
+			var/turf/pos = get_turf(H)
+			pos.add_vomit_floor(H,1) //toxic vomit, for the extra "disgusting" feeling
```
Comment (KazeEspada) [inline code/modules/mob/living/carbon/human/species_types.dm]: Theres a vomit proc for mobs.

Comment (LanCartwright) [inline code/modules/mob/living/carbon/human/species_types.dm]: Wilco. I'll make adjustments.

**Comment by WJohn** [comment]
oh my god this is so disgusting :+1: x :100:

File: code/game/atoms.dm (hunk 1 above)
Comment (RemieRichards) [inline code/game/atoms.dm]: also the reagents list only contains... reagents. so this should be an istypeless for loop.

**Comment by bgobandit** [comment]
now do lizards

**Comment by Cheridan** [comment]
Can I eat vomit as a human, too? Y'know, if I feel like it...

**Comment by CosmicScientist** [comment]
@Cheridan couldn't you scoop it up in a beaker and put it in a drinking glass? Or is that just blood?

**Comment by LanCartwright** [comment]
man you'll sicken me

I'll see what I can do tonight

God damn spaceman sucking puke chunks off the floor.
Absolutely disgusting

Next thing you'll do is make a reagent for vomit and spray that shit everywhere

Well, a guys gotta eat. Even if it's puke chunks.

**Comment by LanCartwright** [comment]
@bgobandit what do you want me to do to lizards?
on my server, they're are unable to get warmer without sleeping. If they get too cold, they instantly fall asleep.

**Comment by KazeEspada** [comment]
I liked the resistant to heat, sleeping when cold pr lizards got. But everyone hated it. Lizard haters thought an extra laser was OP(basically ignoring half the pr). Lizard players hated it because cold making them fall asleep was a huge nerf.

**Comment by Ultimate-Chimera** [comment]
Ew! Ew! https://www.youtube.com/watch?v=7f1MumAJTLo

**Comment by Cheridan** [comment]
inconsistent indentation

**Comment by Zilenan91** [comment]
Conflict fix when this is gonna be awesome

File: code/game/atoms.dm [hunk 4]
```diff
@@ -381,3 +381,25 @@ var/list/blood_splatter_icons = list()
...
+				// make a new proc that does everything below.
+						M.reagents.trans_to(this, M.reagents.total_volume / 10)
+						for(var/datum/reagent/R in reagents.reagent_list)                //clears the stomach of anything that might be digested as food
+							if(istype(R, /datum/reagent/consumable))
+								var/datum/reagent/consumable/nutri_check = R
+								if(nutri_check.nutriment_factor >0)
+									reagents.add_reagent(R.id,R.volume)
+									M.reagents.remove_reagent(R.id,R.volume)
+
+/datum/reagent/proc/lose_lunch(var/atom/splat)
```
Comment (optimumtact) [inline code/game/atoms.dm]: This proc isn't even called by anything. literally has no purpose, it's just code bloat.

**Comment by LanCartwright** [comment]
Massive thanks to oranges for his incredible support. hero of the code. 

I'm losing it. it's 1 am. I'm hearing things.

**Comment by Supermichael777** [comment]
we alredy have grayyyyys who telepathy. They steal hearts.

**Comment by LanCartwright** [comment]
I mean roundstart grayyyyys that can telepathy with other humans.

File: code/game/objects/effects/decals/Cleanable/misc.dm [hunk 5]
```diff
@@ -87,6 +87,22 @@
...
+					if (istype(R, /datum/reagent/consumable))
+						var/datum/reagent/consumable/nutri_check = R
+						if(nutri_check.nutriment_factor >0)
+							H.nutrition += nutri_check.nutriment_factor * nutri_check.volume
+							reagents.remove_reagent(nutri_check.id,nutri_check.volume)
+			reagents.trans_to(H, reagents.total_volume)
+			Destroy()
```
Comment (tkdrg) [inline code/game/objects/effects/decals/Cleanable/misc.dm]: should be `qdel(src)`

**Comment by Zilenan91** [comment]
Oh fuck yeah now I can be disgusting and drink puddles of vomit

File: code/game/atoms.dm [hunk 6]
```diff
@@ -381,3 +367,20 @@ var/list/blood_splatter_icons = list()
...
+		var/obj/effect/decal/cleanable/vomit/V = PoolOrNew(/obj/effect/decal/cleanable/vomit, src)
+		// Make toxins vomit look different
+		if(toxvomit)
+			V.icon_state = "vomittox_[pick(1,4)]"
+		if(M.reagents)
+			clear_reagents_to_vomit_pool(M,V)
+
+/atom/proc/clear_reagents_to_vomit_pool(mob/living/carbon/M, obj/effect/decal/cleanable/vomit/V)
+	M.reagents.trans_to(V, M.reagents.total_volume / 10)
+	for(var/datum/reagent/R in reagents.reagent_list)                //clears the stomach of anything that might be digested as food
```
Comment (phil235) [inline code/game/atoms.dm]: M.reagents , not reagents.

Comment (LanCartwright) [inline code/game/atoms.dm]: I don't know anything about runtimes. This was coded with the help of oranges.

Comment (phil235) [inline code/game/atoms.dm]: When a floor turf tries to call clear_reagents_to_vomit_pool, it encounters an error that stops the proc. The turf's "reagents" var is null, so when you ask it to do use "reagents.reagent_list" and so it shits itself, figuratively.

**Pull request merged**